core/api/src/mill/api/Result.scala

```diff
@@ -1,138 +1,60 @@
-      extends java.lang.Exception(throwable) with Failing[Nothing] {
-    def map[V](f: Nothing => V): Exception = this
-    def flatMap[V](f: Nothing => Result[V]): Exception = this
-
-    override def toString(): String = {
-      var current = List(throwable)
-      while (current.head.getCause != null) {
-        current = current.head.getCause :: current
+    def flatMap[V](f: T => Result[V]): Result[V] = f(value)
+    def get = value
+    def toOption: Option[T] = Some(value)
+    def toEither: Either[String, T] = Right(value)
+    def errorOpt: Option[String] = None
+  }
+  case class Failure(error: String) extends Result[Nothing] {
+    def map[V](f: Nothing => V): Result[Nothing] = this
+
+    def flatMap[V](f: Nothing => Result[V]): Result[Nothing] = this
+    def get = sys.error(error)
+    def toOption: Option[Nothing] = None
+    def toEither: Either[String, Nothing] = Left(error)
+    def errorOpt: Option[String] = Some(error)
+  }
+
+  def fromEither[T](either: Either[String, T]) = either match {
+    case Left(err) => Result.Failure(err)
+    case Right(value) => Result.Success(value)
+  }
+
+  // implementation similar to scala.concurrent.Future#sequence
+  def sequence[B, M[X] <: IterableOnce[X]](in: M[Result[B]])(
+      implicit cbf: BuildFrom[M[Result[B]], B, M[B]]
+  ): Result[M[B]] = {
+    in.iterator
+      .foldLeft[Result[mutable.Builder[B, M[B]]]](Success(cbf.newBuilder(in))) {
+        case (acc, el) =>
+          for (a <- acc; e <- el) yield a += e
       }
-      current.reverse
-        .flatMap { ex =>
-          val elements = ex.getStackTrace.dropRight(outerStack.value.length)
-          val formatted =
-            // for some reason .map without the explicit ArrayOps conversion doesn't work,
-            // and results in `Result[String]` instead of `Array[String]`
-            new scala.collection.ArrayOps(elements).map("    " + _)
-          Seq(ex.toString) ++ formatted
-
-        }
-        .mkString("\n")
-    }
+      .map(_.result())
   }
 
-  class OuterStack(val value: Seq[StackTraceElement]) {
-    def this(value: Array[StackTraceElement]) = this(value.toIndexedSeq)
-
-    override def hashCode(): Int = value.hashCode()
-
-    override def equals(obj: scala.Any): Boolean = obj match {
-      case o: OuterStack => value.equals(o.value)
-      case _ => false
-    }
-  }
-
-  def catchWrapException[T](t: => T): Either[String, T] = {
-    try Right(t)
-    catch {
-      case e: InvocationTargetException =>
-        makeResultException(e.getCause, new java.lang.Exception())
-      case e: Exception => makeResultException(e, new java.lang.Exception())
-    }
-  }
-
-  def makeResultException(e: Throwable, base: java.lang.Exception): Left[String, Nothing] = {
-    val outerStack = new mill.api.Result.OuterStack(base.getStackTrace)
-    Left(mill.api.Result.Exception(e, outerStack).toString)
-  }
+  class Exception(val error: String) extends java.lang.Exception(error)
```

Comment:
Some ScalaDoc would be nice. What's the purpose of and when to use this exception?

Reply:
This is meant to replace the `Result.Failure extends Exception` we wanted to get rid of. I'll add Scaladoc later, I expect this to evolve significantly